website/content/en/highlights/2022-05-03-0-22-0-upgrade-guide.md

```diff
@@ -115,3 +116,56 @@ You also still have the option to use raw strings (`s'...'`):
 ```coffee
 assert!("\{{ right here \}}" == s'{{ right here }}')
 ```
+
+#### VRL's `parse_syslog` now nests structured data fields {#vrl-parse_syslog}
+
+The `parse_syslog` VRL function now returns the structured data elements as a nested map under the
+structured data element name, rather than prefixing each key with the element name, to match the
+behavior of the `syslog` source.
+
+That is:
+
+```
+parse_syslog!(s'<1>1 2022-04-25T23:21:45.715740Z Gregorys-MacBook-Pro.local 2d4d9490-794a-4e60-814c-5597bd5b7b7d 79978 - [exampleSDID@32473 foo.baz="bar"] test message')
+```
+
+Now returns:
+
+```json
+{
+  "appname": "2d4d9490-794a-4e60-814c-5597bd5b7b7d",
+  "exampleSDID@32473": {
+    "foo.baz": "bar"
+  },
```

Comment:
Aaah, yes. I think _that_ might actually undesired behavior 🤔  I think the nested keys should just be inserted flat since syslog's structured data is simply key/value.

Reply:
Umm yeah maybe that's a good compromise.
Another good argument for that idea of setting a "limit" of depth=1 for nesting is to protect Vector from blowing up with malicious syslog data, such as:
```
<1>1 2022-04-25T23:21:45.715740Z Gregorys-MacBook-Pro.local 2d4d9490-794a-4e60-814c-5597bd5b7b7d 79978 - [exampleSDID@32473 a.b.c.d.e.f.g.h.i.j.k.l.m.n.o.p.q.r.s.t.u.v.w.x.y.z="bar"] test message
```
(you get the idea).

The `syslog` source should be modified to do the same then to be aligned with the VRL function.

Reply:
👍 I opened [11a61b0](https://github.com/vectordotdev/vector/pull/12433/commits/11a61b02ed086b1bccc407a9884b3e35c4a2f033) to bring the `syslog` handling of structured data fields that contain a `.` in line.

Reply:
Your implementation got me curious to go check the spec for repeated key names in the same SD element, a-la HTTP query parameters that can be repeated and thus requiring a multimap.

The sad news are that they CAN be repeated: https://datatracker.ietf.org/doc/html/rfc5424#section-6.3.3
> An SD-PARAM MAY be repeated multiple times inside an SD-ELEMENT.

And Vector's will silently ignore them:
```
$ parse_syslog!(s'<1>1 2022-04-25T23:21:45.715740Z Gregorys-MacBook-Pro.local 2d4d9490-794a-4e60-814c-5597bd5b7b7d 79978 - [exampleSDID@32473 a="bar" a="foo"] test message')
{ "appname": "2d4d9490-794a-4e60-814c-5597bd5b7b7d", "exampleSDID@32473.a": "foo", "facility": "kern", "hostname": "Gregorys-MacBook-Pro.local", "message": "test message", "procid": 79978, "severity": "alert", "timestamp": t'2022-04-25T23:21:45.715740Z', "version": 1 }
```

The good news is that I think it should be addressed in a separate issue/PR :)

The funny news is that this would be yet another breaking change to address because the data model will need to accomodate for this, similar to #11576 :trollface: 

... don't hate me guys :(

Reply:
🤣 nice find. Agreed on both counts: we should parse that into an array if the same key is seen multiple times and we can handle that as a separate follow-up.

Reply:
Opened https://github.com/vectordotdev/vector/issues/12435 to track that.